qlib/rl/order_execution/strategy.py

```diff
@@ -0,0 +1,143 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+import collections
+from abc import ABCMeta
+from typing import Any, cast, Dict, Generator
+
+import pandas as pd
+from qlib.backtest import CommonInfrastructure, Order
+from qlib.backtest.decision import BaseTradeDecision, TradeDecisionWO, TradeRange, TradeRangeByTime
+from qlib.backtest.utils import LevelInfrastructure, SAOE_DATA_KEY
+from qlib.rl.data.exchange_wrapper import QlibIntradayBacktestData
+from qlib.rl.order_execution.state import QlibBacktestAdapter, SAOEState
+from qlib.rl.order_execution.utils import get_ticks_slice
+from qlib.strategy.base import RLStrategy
+
+
+class SAOEStrategy(RLStrategy, metaclass=ABCMeta):
+    """RL-based strategies that use SAOEState as state."""
+
+    def __init__(
+        self,
+        policy: object,  # TODO: add accurate typehint later.
+        outer_trade_decision: BaseTradeDecision = None,
+        level_infra: LevelInfrastructure = None,
+        common_infra: CommonInfrastructure = None,
+        **kwargs: Any,
+    ) -> None:
+        super(SAOEStrategy, self).__init__(policy, outer_trade_decision, level_infra, common_infra, **kwargs)
+
+        self.adapter_dict: Dict[tuple, QlibBacktestAdapter] = {}
+
+    def _create_qlib_backtest_adapter(self, order: Order, trade_range: TradeRange) -> QlibBacktestAdapter:
+        if not self.common_infra.has(SAOE_DATA_KEY):
+            self.common_infra.reset_infra(**{SAOE_DATA_KEY: {}})
+
+        # saoe_data can be considered as some type of cache. Use it to avoid unnecessary data reload.
+        # The data for one order would be loaded only once. All strategies will reuse this data.
+        saoe_data = self.common_infra.get(SAOE_DATA_KEY)
+        if order.key not in saoe_data:
+            data = self.trade_exchange.get_deal_price(
+                stock_id=order.stock_id,
+                start_time=order.start_time.replace(hour=0, minute=0, second=0),
+                end_time=order.start_time.replace(hour=23, minute=59, second=59),
+                direction=order.direction,
+                method=None,
+            )
+
+            ticks_index = pd.DatetimeIndex(data.index)
+            if isinstance(trade_range, TradeRangeByTime):
+                ticks_for_order = get_ticks_slice(
+                    ticks_index,
+                    trade_range.start_time,
+                    trade_range.end_time,
+                    include_end=True,
+                )
+            else:
+                ticks_for_order = None  # FIXME: implement this logic
+
+            backtest_data = QlibIntradayBacktestData(
+                order=order,
+                exchange=self.trade_exchange,
+                start_time=ticks_for_order[0],
+                end_time=ticks_for_order[-1],
+            )
+
+            saoe_data[order.key] = (ticks_index, ticks_for_order, backtest_data)
```

Comment:
Would `saoe_data` get larger and larger and overwhelm the memory?

Reply:
That's a good point. I use "LRUCache" to store this data now.